### ISSUE TITLE
Possible to continue setting data after building index?

### ISSUE THREAD

**Issue by FangLinHe** [searchivarius/nmslib#73]
Hello,

I really appreciate your great work, which is really efficient and high quality!

However, I am encountering a problem: since my dataset is big (more than 1 million features with 4096 dimension), I would like to build the index only one time, and when the new dataset comes in, I could simply call `nmslib.setData(self.index, i, feature)` where i would be greater than `n` specified during initialization `index = nmslib.initIndex(n, ...)`.  Is it possible to do that in some way?  Thank you very much!

**Comment by searchivarius** [comment]
Hi, thank you for the kind words. 

Are you asking whether it is possible
1) save/restore index
2) add more data after the index is created (i.e., incremental updates).

**Comment by FangLinHe** [comment]
Hello searchivarius,

Thank you for your prompt reply! By that I meant whether it is possible to 2) add more data after the index is created (i.e., incremental updates).

As what I have read in the manual, it seems that some methods have to be re-indexed for new coming data, e.g. small-world graph. But for projection-based methods, I believe it is possible to achieve the goal of incremental updates.

**Comment by searchivarius** [comment]
@FangLinHe, interestingly sw-graph index *IS* created incrementally. This is also true for some other methods. Yet, the generic API doesn't allow you to do so. Do you play to use the code from Python?
BTW, we have a branch where two most important indices can be read/written. But, it is not well tested yet.

**Comment by FangLinHe** [comment]
@searchivarius Oh, that is my fault to misunderstood.  I am actually a newbie at this area.
Yes, I am using Python API; so you may mean that, in C++ it allows me to incrementally update the index?  It would be great if you could give me some hint how to achieve it by modifying the Python API, or even better if you also plan to do so in the future release. :)
And great to know that you have a branch being able to read/write indices.  I would love to give it a try in the future.

**Comment by searchivarius** [comment]
@FangLinHe , eehh, actually it wouldn't be possible without rewriting the code quite a bit. I think it's a shame we can't support it yet, given that many method, in principle, can do it easily. I have added the issue to the release 2.0.

**Comment by FangLinHe** [comment]
@searchivarius So sad to hear that, but you really did a good job.  I will still use it and wait for the release 2.0 to support incremental update!  Thank you so much for the discussion and answering my doubts! :)

**Comment by searchivarius** [comment]
Please, don't close the issue though :-)

**Comment by FangLinHe** [comment]
Oops, my fault. Sorry!

**Comment by FangLinHe** [comment]
Reopen

**Comment by FangLinHe** [comment]
Reopen

**Comment by funtusov** [comment]
@searchivarius I am sorry, it was a bit unclear, could you please clarify, whether appending data to an existing index isn't easily achievable only in the Python API or with C++ directly as well?

**Comment by searchivarius** [comment]
@funtusov Currently, we support only fully static indices, sorry. At the high level, it isn't hard to make extendible data sets, but this would require change quite a few things all the way down.

**Comment by ajkl** [comment]
Looking forward to this feature.

**Comment by searchivarius** [comment]
I have recently added  support for incremental addition and/or deletion (all should be done in batches), but it requires using C++ directly.

**Comment by searchivarius** [comment]
I have recently added  support for incremental addition and/or deletion (all should be done in batches), but it requires using C++ directly.

**Comment by leafjungle** [comment]
@searchivarius ， Great.  is there any example to add new data to exist index by c++?

**Comment by searchivarius** [comment]
Hi @leafjungle here it is: https://github.com/searchivarius/nmslib/blob/master/test_batch_app/test_batch_mod.cc
HNSW, unfortunately, doesn't support batch deletion and addition yet, only SW-graph.

**Comment by searchivarius** [comment]
@leafjungle not HNSW, SW-graph only. Making HNSW dynamic isn't impossible, but I haven't done this yet.

**Comment by mgwizdala** [comment]
@searchivarius but even for sw-graph, it seems to be impossible to add new point batch from python binding without calling createIndex. 
Any idea how to do this without rebuilding the whole index? (again from python binding)

**Comment by searchivarius** [comment]
@mgwizdala this is correct, Python bindings don't support this.

**Comment by baklazan** [comment]
Hello,

I'm currently trying to implement dynamic insertion in HNSW. After reading the code of class `Hnsw<dist_t>` a little, I have following question:

What is the purpose of `data_rearranged_`? Does `data_rearranged[i]` differ from `ElList_[i]->getData()`? 

I would like to get rid of `data_rearranged_` as it prevents me from relocating `data_level0_memory` (`data_rearranged_` becomes invalid after relocation of `data_level0_memory`).

Sorry if this is not the correct place to ask this question.

**Comment by searchivarius** [comment]
@baklazan there are two types of indices. One that uses data_ only. It is actually dynamic already. You can add elements to data_ and just call the add() function. It's just not visible to outside API. Buuut, there is also an optimized fully-static version of the index, which uses on the data_rerranged. If you set skip_optimized_index flag to 1, this static version of the index is never created.

**Comment by baklazan** [comment]
>  You can add elements to data_ and just call the add() function.

Yeah, I already did that (I also resized `VisitedList`s in `visitedlistpool`).
But I'd like to make the optimized index dynamic as well, if possible.

**Comment by searchivarius** [comment]
@baklazan it's ether static and fast or dynamic and somewhat slower. To update the static index incrementally, you have to implement some sort of an efficient merge. I wouldn't bother about this. I think people who need truly and fully dynamic indices can bite the bullet and use a 20-30% slower version of HNSW.

If the retrieval speed is paramount, but you still need to update the data, you would just use a combination of the fast but large static index and a small dynamic one. Periodically, the static index would need to be rebuilt to incorporate updates.

**Comment by baklazan** [comment]
@searchivarius Ok, thank you.

**Comment by phdowling** [comment]
Are there plans to add this to the Python bindings? Real-time index additions are super valuable. And I'm not sure if I understand correctly: currently only SW-graph supports dynamic indexing, and HNSW _might_ support this later on?

Actually, I also just came across this: https://github.com/nmslib/hnsw/issues/2 - is this the same in nmslib, or is the dynamic updating only supported by the implementation in that repository?

**Comment by yurymalkov** [comment]
@phdowling 
>Actually, I also just came across this: nmslib/hnsw#2 - is this the same in nmslib, or is the dynamic updating only supported by the implementation in that repository?

It is a reimplementation, optimized for less memory usage, faster construction and incremental indexing. 
Nmslib's c++ interface also supports incremental insertions for sw-graph and hnsw (only regular indexes).
If you only need regular dense spaces with incremental updates I would advice using hnswlib.

**Comment by AVBelyy** [comment]
Hello @searchivarius and @yurymalkov and thanks for the great library!
We tried and tested your library with hnsw and sw-graph methods on large scale text classification sparse datasets (# objects ~ 2 millions, # features ~ 1 million) are were impressed with the results! We would now like to use your method to build efficient SVM multi-class classification iterative method.

Inside this method, on each iteration, we will perform a series of updates (i.e. deletion-insertions) to  few weight vectors (i.e. vectors in hnsw index). There are few questions to this approach:
1) Do you think the performance of the index will degrade significantly if we do a large series of incremental batch updates (like, 100k or 1m updates for index with ~1m vectors)?
2) Do I understand correctly that in order to experiment with batch insertions and deletions in Python, we will have to implement Python bindings ourselves? Is it at all possible and how much time/effort do you think it would take?

Thanks a lot in advance and thanks for the library again!

**Comment by searchivarius** [comment]
Hi @AVBelyy https://github.com/nmslib/hnsw supports deletions and insertions, I think. For SW-graph I experimented with deletions and insertions. Performance doesn't seem to deteriorate quickly: I think I have done something like 100K updates for 1M dataset and it was fine. However, not sure about 1M. Depends on updates.

**Comment by AVBelyy** [comment]
I've checked with implementation code and hnsw seems to only support insertions for now. So SW-graph looks like the only choice for performing incremental updates.
And what about Python bindings for batch updates: should I expect any difficulty implementing it? I can do it & send a PR after it's done.

**Comment by searchivarius** [comment]
@AVBelyy ohh, right, sorry I was convinced it already has deletions. Well, if static SW-graph performance is good enough for you, I can certainly guide you with respect to embeddings. I will also create a new dynamic version of nmslib, but it will take time.

**Comment by 8W9aG** [comment]
I was wondering if this was implemented on the latest version? It's a little unclear to me if `addDataPoint` et al allows for dynamic additions or just additions before building the index.

**Comment by searchivarius** [comment]
@8W9aG No, sorry. It would require reworking the library a lot.

**Comment by 8W9aG** [comment]
@searchivarius I saw you posted this PR for support of dynamic vectors in SW https://github.com/nmslib/nmslib/blob/master/test_batch_app/test_batch_mod.cc and was wondering how this worked? Is `AddBatch` in this case accepting vectors after the index has been created?

**Comment by searchivarius** [comment]
@8W9aG it's just implemented internally for one method without exposing this to Python.

**Comment by 8W9aG** [comment]
@searchivarius Can you tell me what would be needed to bring this to the python bindings? As far as I can tell this is how I would attempt this:

The index class has new methods such as `addDynamic` that allows adding vectors. When adding vectors either the indexing library has support for this, or the vectors distance from the search is always computed (eventually this would have a hit on performance, but I think this is fine as long as it is clear in the docs). The new vectors are stored in a new `ObjectVector` instance on the index and the index of the new vectors is the indexed size + the index of the result in the `ObjectVector`.

Would this be amenable? I was pretty much going to do this in a wrapper around nmslib but I figured that I may as well push this functionality into the library itself, since its a problem I find myself having frequently.

**Comment by searchivarius** [comment]
The first thing to do is to think how this can be done in the generic fashion. I would ideally retain the interface when the library is updated and more methods support dynamic updates.

**Comment by 8W9aG** [comment]
I think in order to do this in a generic fashion, use the Index classes virtual methods to implement naive dynamic updates (brute force), and if the implementation supports smarter dynamic updates, add those dynamic methods in the concrete class itself, what do you think?

**Comment by searchivarius** [comment]
@8W9aG I see, but now you have coupled indexing and data-addition procedures.

**Comment by 8W9aG** [comment]
That is true, but I submit to you that the dynamic updates are still part of the index and should be included in that class, despite the fact that they may be slower to iterate over. If this is covered in the documentation perhaps that is OK?

**Comment by searchivarius** [comment]
@8W9aG I worry mostly about the interface compatibility and sensibility part. E.g., in SQL you create a table and an index. Then, additions are indexed automatically. Here, we call addData* and then we call index. These are two separate calls.

**Comment by 8W9aG** [comment]
Do you have an idea how this API would ideally look?

**Comment by searchivarius** [comment]
@8W9aG sorry fo the delay, I had to go through the code. We do not have much choice and we need to mimic existing API used in https://github.com/nmslib/nmslib/blob/master/test_batch_app/test_batch_mod.cc
So, basically, one needs to support functions ``addBatch`` and ``deleteBatch`` at the level of the Index/Search Method class, but here's one big catch related to the lifecycle of data points/objects. The index, currently, and this is a design flaw, cannot on its own handle data. Data needs to be handled externally. So, whenever you call ``addDataPoints`` you need to memorize the data points you added somewhere. You add a batch, you need to memorize the data points somewhere and then call the indexing function. 

Luckily the Python bindings class holds both the index and data, so this drawback can be overcome.

However, things are still a bit non-trivial:
1. Currently addDataPointBatch simply replaces all existing data points. It should probably extend them.
2. How do we deal with the existing createIndex method? Do we add an updateIndex method that would accept a batch of somehow specified data points?
3. On a related notes, to maintain a similarity, with the createIndex method, should we call addDataPointBatch first and then updateIndex second? Or do we want to combine this in a single call?
4. A similar question is for the deleteBatchFunction.
5. BTW, for a dynamic small world the save/load function would need to be modified a bit, but this is going to be a relatively minor change. Again, the current difficulty is that a search method cannot handle data on its own. Perhaps, this **must be changed** before we make changes to the Python API.

**Comment by searchivarius** [comment]
PS: that index doesn't handle data isn't totally a design flaw, actually, it was sort of a convenient thing for testing purposes: data can be resplit easily.

**Comment by tlyalex** [comment]
> @phdowling
> 
> > Actually, I also just came across this: [nmslib/hnswlib#2](https://github.com/nmslib/hnswlib/issues/2) - is this the same in nmslib, or is the dynamic updating only supported by the implementation in that repository?
> 
> It is a reimplementation, optimized for less memory usage, faster construction and incremental indexing.
> Nmslib's c++ interface also supports incremental insertions for sw-graph and hnsw (only regular indexes).
> If you only need regular dense spaces with incremental updates I would advice using hnswlib.


Is this the function that could be used to add new point into the regular hnsw index?
https://github.com/nmslib/nmslib/blob/ab83539b4f8f56d6f6ad4242838e61c398815bbc/similarity_search/src/method/hnsw.cc#L502

Thanks!

**Comment by searchivarius** [comment]
@tlyalex in principle, yes, but the dataset array has to be expanded as well.

**Comment by searchivarius** [comment]
PS: @tlyalex in NMSLIB this would work only for the so-called non-optimized indices.
